cpp/src/arrow/filesystem/filesystem_utils.cc

```diff
@@ -0,0 +1,286 @@
+    auto search = options.find("use_hdfs3");
+    if (search != options.end()) {
+      if (search->second.compare("1") == 0) {
+        return true;
+      } else {
+        return false;
+      }
+    } else {
+      return false;
+    }
+  }
+
+  int GetRepsNum() {
+    auto search = options.find("replication");
+    if (search != options.end()) {
+      if (search->second.empty()) {
+        return 3;
+      } else {
+        return std::stoi(search->second);
+      }
+    } else {
+      return 3;
+    }
+  }
+
+ private:
+  FileSystemType GetFileSystemTypeFromString(const std::string& s) {
+    if (s == "hdfs") {
+      return FileSystemType::HDFS;
+    }
+    if (s == "http") {
+      return FileSystemType::S3;
+    }
+    if (s == "https") {
+      return FileSystemType::S3;
+    }
+    if (s == "file") {
+      return FileSystemType::LOCAL;
+    }
+    if (s.empty()) {
+      return FileSystemType::LOCAL;
+    }
+    return FileSystemType::UNKNOWN;
+  }
+
+  Status ParseURI(const std::string& s) {
+    arrow::internal::Uri uri;
+    RETURN_NOT_OK(uri.Parse(s));
+    fs_type_ = GetFileSystemTypeFromString(uri.scheme());
+    switch (fs_type_) {
+      case FileSystemType::HDFS:
+        host_ = uri.host();
+        port_ = uri.port();
+        RETURN_NOT_OK(ParsePath(uri.path()));
+        RETURN_NOT_OK(ParseOptionsFromPath(s));
+        break;
+      case FileSystemType::LOCAL:
+        RETURN_NOT_OK(ParsePath(uri.path()));
+        break;
+      case FileSystemType::S3:
+        return Status::NotImplemented("S3 is not supported yet.");
+      default:
+        break;
+    }
+    return Status::OK();
+  }
+
+  Status ParsePath(const std::string& path) {
+    auto dir_file = internal::GetAbstractPathParent(path);
+    directory_ = dir_file.first;
+    file_name_ = dir_file.second;
+    if (directory_.empty() || file_name_.empty()) {
+      return Status::Invalid("Invalid hdfs path");
+    }
+    return Status::OK();
+  }
+
+  Status ParseOptionsFromPath(const std::string& s) {
+    std::string search_str = std::string("?");
+    std::string::size_type pos = s.find_first_of(search_str);
+    if (pos != std::string::npos) {
+      RETURN_NOT_OK(ParseOptions(s.substr(pos + 1)));
+    }
+    return Status::OK();
+  }
+
+  Status ParseOptions(const std::string& s) {
+    std::string::size_type pos2, pos0 = 0, pos1 = 0;
+    std::string search_str0 = std::string("&");
+    std::string search_str1 = std::string("=");
+    std::string key, value;
+    while (pos0 < s.size()) {
+      pos1 = s.find_first_of(search_str0, pos0);
+      if (pos1 == std::string::npos) {
+        pos1 = s.size();
+      }
+      auto matched = s.substr(pos0, pos1 - pos0);
+      pos2 = matched.find_first_of(search_str1);
+      key = matched.substr(0, pos2);
+      if (pos2 == pos1 - pos0) {
+        value = "";
+      } else {
+        value = matched.substr(pos2 + 1);
+      }
+      options.emplace(std::make_pair(std::string(key), std::string(value)));
```

Comment:
Emplace constructs a pair implicitly:
```suggestion
      options.emplace(std::move(key), std::move(value));
```